### PR TITLE
Issue #2 Completed

### PR DESCRIPTION
**Pull request by farshidgoudarzi**
**Feature - Implement a read status for messages #2**
I used a timestamp for when messages of a conversation was viewed by the user. So basically any messages after that would be unread. Works if user is viewing the conversation and messages come (Conversation is updated), and changes are applied in realtime.
Screenshot of the UI are attached as requested.

Let me know what you think.

Issue2_SS

### CODE REVIEW

**Comment by farshidgoudarzi** [comment]
There was a conflict since I had started with the second issue before we close the first, so I rebased it on top of main.

**Comment by farshidgoudarzi** [comment]
Another point, the issue could be resolved by adding a viewed flag to messages also. But since in the next issue, there is the mention of changing the structure so that a conversation could have multiple participants, that approach would not be efficient there (a message would have multiple viewers), that is why I went with the last viewed timestamp. And overall this one seems better I think. Let me know what you think.

File: server/db/models/conversation.js
Comment (1234tgk) [inline server/db/models/conversation.js]: This is one way to implement this feature. This allows potentially multiple users within the conversation. However, dealing with time is very finicky, especially for asynchronous codes (even right now, the app has a bit delay on updating the read messages when I tested with GitPod).

The simplest way to implement this feature is, as you mentioned, is assigning read key to each message. In terms of multiple users, I believe there is a way to implement this even with multiple users (maybe you can think about it while doing #3. It is not a requirement however 🙂)

Can you try implement this feature by assigning the read key to each messages instead, please? I know this is a lot, but much of the logics are there, and you just need to tweak here and there.

File: server/db/models/conversation.js
Comment (1234tgk) [inline server/db/models/conversation.js]: You should just throw an error here 🙂 

```suggestion
  if (!conversation) throw new Error('conversationId not valid');
```

File: server/routes/api/conversations.js
Comment (1234tgk) [inline server/routes/api/conversations.js]: Several things here:

1. Since you are marking the messages as read (and not creating anything new in the database), this should be either 'PUT' or 'PATCH' request route.
2. The name of the REST API should be noun (such as messages) instead of verb or action (i.e. mark-as-read).
3. The naming convention for routes is kebab-case, not camelCase 🙂

File: server/routes/api/conversations.js
Comment (1234tgk) [inline server/routes/api/conversations.js:11]: Nice catch here! However, you also want to protect the route from users not part of the conversation by returning early with the appropriate status code.

File: client/src/store/utils/thunkCreators.js
Comment (1234tgk) [inline client/src/store/utils/thunkCreators.js]: Please remove any console logs you wrote for debugging purposes since they hinder codes readability.

File: client/src/store/conversations.js
Comment (1234tgk) [inline client/src/store/conversations.js]: Same here 🙂

File: client/src/components/Sidebar/ChatContent.js
Comment (1234tgk) [inline client/src/components/Sidebar/ChatContent.js]: Doing this inside the frontend reduces the performance in the client side. You can instead find out how much unread messages for certain user with Sequelize's [count](https://sequelize.org/master/manual/model-querying-basics.html#-code-count--code-) method.

File: client/src/components/ActiveChat/ActiveChat.js
Comment (1234tgk) [inline client/src/components/ActiveChat/ActiveChat.js]: Same here 🙂

**Review by 1234tgk — COMMENTED** [review]
Great work on #2. Your feature works well minus the fact that it has tiny bit of lag in terms of the notification. There are several points I want to make:

1. Ideally you want to implement this feature by putting read key to each message. More on this on my comments.
2. In the sidebar, if the last message is unread, the message should be displayed bolded.
3. There should be a small image of the receiver of the message under the last message read in the conversation itself, like shown in the mocks. This avatar image should also update in real time.

Please address these issues before requesting for the second review. Let us know down in the Slack if you have any questions too. 😄

File: client/src/store/utils/thunkCreators.js
Comment (farshidgoudarzi) [reply client/src/store/utils/thunkCreators.js]: Sure will do, I think a few got away from me.

File: server/db/models/conversation.js
Comment (farshidgoudarzi) [reply server/db/models/conversation.js]: Ok, yes I missed the point that the other user must see the last message that is read. In that case I need to implement the read flag for the messages. Since I wasn't implementing the whole UI I didn't get the objective from the mockup. You are right I will get to it.

File: server/routes/api/conversations.js
Comment (farshidgoudarzi) [reply server/routes/api/conversations.js]: Good point, will change it.

File: server/routes/api/conversations.js
Comment (farshidgoudarzi) [reply server/routes/api/conversations.js:11]: Ok, should we send 401 if the user is not in the conversation? I'm not sure if it counts as an Auth error...

File: client/src/components/Sidebar/ChatContent.js
Comment (farshidgoudarzi) [reply client/src/components/Sidebar/ChatContent.js]: Good point. Sure I will do it with count. Seems a lot better.

**Comment by farshidgoudarzi** [comment]
I remade the solution with the read flag on each message. All of the above points are addressed. I also attached new UI screen shot.

Just to point out, I found some bugs in the app, specially in sending the events on the socket. For example when sending message on the new chat, if another user is connected, they will get the new message as if it was theirs, even though it was sent to another user (recipient is not being filtered)... But since it was not mentioned in the issues, and we are not working on a production project, I didn't put time one them.

Also the issue with the lag was for sending the event on channel... I don't think using the timestamp flag would actually be slower because it is the same principle, but I the solution with the messages flag as request.

Let me know what you think.

Issue2_SS

File: client/src/components/ActiveChat/ActiveChat.js
Comment (FilipeGorodscy) [inline client/src/components/ActiveChat/ActiveChat.js:34]: I don't think you need to loop here to retrieve the `otherMessages` array. You can simply update the database using the `conversationId` 👍

File: client/src/components/Sidebar/ChatContent.js
Comment (FilipeGorodscy) [inline client/src/components/Sidebar/ChatContent.js:43]: You should remove any unnecessary commented-out code when making a PR :)

File: client/src/components/Sidebar/ChatContent.js
Comment (FilipeGorodscy) [inline client/src/components/Sidebar/ChatContent.js:47]: You can pass `props` to the `useStyles` hook :)

File: server/db/models/message.js
Comment (FilipeGorodscy) [inline server/db/models/message.js:17]: 💯

File: server/routes/api/conversations.js
Comment (FilipeGorodscy) [reply server/routes/api/conversations.js:11]: `403: Forbidden` fits better in this case :)

File: server/routes/api/conversations.js
Comment (FilipeGorodscy) [inline server/routes/api/conversations.js:32]:
```suggestion
        return res.sendStatus(403);
```

**Review by FilipeGorodscy — APPROVED** [review]
Great job overall! Feel free to merge this PR after addressing the feedback below 💯

**Pull request merged**
